### PR TITLE
Pull request for Issue1366: create general pseudomotor class

### PR DESCRIPTION
**Pull request by helfrij**

### CODE REVIEW

**Comment by helfrij** [comment]
The basic AMPseudoMotorControl interface and changes to the classes named in #1366 are ready for first round of review.
@dretrex @XuetaoLiu-CLS @iainworkman @davidChevrier @epengr @dleacock

File: source/analysis/AM1DDarkCurrentCorrectionAB.cpp [hunk 1]
```diff
@@ -213,7 +213,8 @@ bool AM1DDarkCurrentCorrectionAB::values(const AMnDIndex &indexStart, const AMnD
 		if (dwellTimes.at(i) == 0 || dwellTimes.at(i) < 0)
 			return false;
 
-		outputValues[i] = data.at(i)/(dwellTimes.at(i) * timeUnitMultiplier_) - darkCurrent_;
+//		outputValues[i] = data.at(i)/(dwellTimes.at(i) * timeUnitMultiplier_) - darkCurrent_;
```
Comment (dretrex) [inline source/analysis/AM1DDarkCurrentCorrectionAB.cpp]: Commented out code can probably be removed.

File: source/beamline/AMPseudoMotorControl.cpp [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+
+	return AMControl::NoFailure;
+}
+
+bool AMPseudoMotorControl::stop()
+{
+	bool result = true;
+
+	foreach (AMControl *child, children_) {
+		if (child)
```
Comment (dretrex) [inline source/beamline/AMPseudoMotorControl.cpp]: To be thorough, you may want to have this if-statement incorporate a childs ability to stop (i.e.: ```if (child && child->canStop())```).  Food for thought.  That may open up other issues, but it probably makes sense to think about them now.

File: source/beamline/AMPseudoMotorControl.h [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+	virtual double maximumValue() const { return maximumValue_; }
+
+	/// Returns true if this control is connected, false otherwise.
+	virtual bool isConnected() const { return connected_; }
+	/// Returns true if the control is moving.
+	virtual bool isMoving() const { return isMoving_; }
+	/// Returns true if the control is moving, as a result of this control's action.
+	virtual bool moveInProgress() const { return moveInProgress_; }
+
+//	/// Returns true if the control value is always measureable, provided it is connected. False otherwise.
```
Comment (dretrex) [inline source/beamline/AMPseudoMotorControl.h]: You can probably remove this commented out code too.

File: source/beamline/BioXAS/BioXASCarbonFilterFarmActuatorControl.cpp [hunk 4]
```diff
@@ -49,23 +39,58 @@ BioXASCarbonFilterFarmActuatorControl::~BioXASCarbonFilterFarmActuatorControl()
...
+{
+	return BioXASCarbonFilterFarm::Window::Top;
+}
+
+bool BioXASCarbonFilterFarmActuatorControl::canMeasure() const
+{
+	bool result = false;
+
+	if (isConnected())
+		result = position_->canStop();
```
Comment (dretrex) [inline source/beamline/BioXAS/BioXASCarbonFilterFarmActuatorControl.cpp]: Is this supposed to be ```canMeasure()```?

File: source/beamline/BioXAS/BioXASCarbonFilterFarmActuatorControl.h [hunk 5]
```diff
@@ -38,100 +35,73 @@ class BioXASCarbonFilterFarmActuatorControl : public AMCompositeControl
...
-	/// The current move state, true if this control has initiated a move.
-	bool moveInProgress_;
-	/// The mapping between window and position.
-	QMap<Window::Selection, double> positionMap_;
+	/// The mapping between window enum and physical window position.
+	QMap<double, double> positionMap_;
+
 	/// The position control.
-	AMPVControl *position_;
+	AMControl *position_;
```
Comment (dretrex) [inline source/beamline/BioXAS/BioXASCarbonFilterFarmActuatorControl.h]: AMPVControl position + AMReadOnlyControl status => AMPVwStatusControl?

File: source/beamline/BioXAS/BioXASSSRLMonochromatorEnergyControl.cpp [hunk 6]
```diff
@@ -1,9 +1,38 @@
...
+	maximumValue_ = 1000000;
+
+	setAllowsMovesWhileMoving(false);
+	setContextKnownDescription("Energy");
+
+	// Initialize member variables.
+
+	hc_ = 12398.42;
+	crystal2D_ = 3.8403117;
+	m1AngleOffset_ = -0.4;
```
Comment (dretrex) [inline source/beamline/BioXAS/BioXASSSRLMonochromatorEnergyControl.cpp]: This can change can't it?

Comment (dretrex) [inline source/beamline/BioXAS/BioXASSSRLMonochromatorEnergyControl.cpp]: Never mind.

File: source/beamline/BioXAS/BioXASSSRLMonochromatorEnergyControl.h [hunk 7]
```diff
@@ -6,64 +6,130 @@
...
 	virtual ~BioXASSSRLMonochromatorEnergyControl();
 
-	/// Returns true if all subcontrols are connected, false otherwise.
-	virtual bool isConnected() const;
+	/// Returns true if the control value is always measureable, provided it is connected. False otherwise.
+	virtual bool shouldMeasure() { return true; }
+	/// Returns true if a control move is always possible, provided it is connected. False otherwise.
+	virtual bool shouldMove() const { return true; }
+	/// Returns true if a control stop is always possible, provided it is connected. False otherwise.
+	virtual bool shouldStop() const { return false; }
```
Comment (dretrex) [inline source/beamline/BioXAS/BioXASSSRLMonochromatorEnergyControl.h]: Is there a reason why the mono shouldn't be able to stop?

File: source/beamline/BioXAS/BioXASCarbonFilterFarmActuatorControl.h (hunk 5 above)
Comment (helfrij) [inline source/beamline/BioXAS/BioXASCarbonFilterFarmActuatorControl.h]: I think it's best to keep them separate for now, as the status pv reports whether or not the position control is in a valid position (not it's moving status).

File: source/beamline/CLS/CLSJJSlitBladesControl.h [hunk 8]
```diff
@@ -3,66 +3,53 @@
...
 
-#include "beamline/AMCompositeControl.h"
+#include "beamline/AMPseudoMotorControl.h"
 #include "actions3/AMAction3.h"
 
 // Values.
 
-#define CLSJJSLITSCONTROL_VALUE_MIN -30
-#define CLSJJSLITSCONTROL_VALUE_MAX 30
+#define CLSJJSLITBLADESCONTROL_VALUE_MIN -30
```
Comment (dretrex) [inline source/beamline/CLS/CLSJJSlitBladesControl.h]: Are these actually hard coded values?  Are they different from using your setMinimumValue and setMaximumValue?

File: source/ui/BioXAS/BioXASSSRLMonochromatorRegionControlEditor.cpp [hunk 9]
```diff
@@ -27,6 +29,8 @@ void BioXASSSRLMonochromatorRegionControlEditor::setControl(AMControl *newContro
 
 void BioXASSSRLMonochromatorRegionControlEditor::onRegionControlMoveStarted()
 {
+	qDebug() << "\nRegion control move started.\n";
```
Comment (dretrex) [inline source/ui/BioXAS/BioXASSSRLMonochromatorRegionControlEditor.cpp]: Probably don't need these qDebug()'s any more.

**Comment by dretrex** [comment]
Overall, this looks good!

**Comment by helfrij** [comment]
I've updated the changes to AMExtendedControlEditor--the widget titles should now display/update appropriately. We have a variety of editors on Side--some with borders/titles hidden, some with titles set, and others with no titles--and they appear to work as expected.

Still to do testing-wise:
- XAS scan with new energy pseudomotor implementation
- region change?

**Comment by helfrij** [comment]
I think this pull request is ready to merge. Testing for some items has been tricky because of scans running on Main and work being done on the Side mono. It may be best to merge and then deal with possible issues as they crop up in separate branches.

**Pull request merged**